In [8]:
import pandas as pd

In [9]:
df = pd.read_csv("OnlineRetail.csv", encoding='latin1')

In [14]:
# First rows
df.head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,12/1/2010 8:26,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,12/1/2010 8:26,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,12/1/2010 8:28,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,12/1/2010 8:28,1.85,17850.0,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,12/1/2010 8:34,1.69,13047.0,United Kingdom


In [13]:
# Shape
df.shape

# Info
df.info()

# Data types
df.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

The dataset contains 541,909 records and 8 columns.
The numerical columns are Quantity, UnitPrice, and CustomerID, while the remaining columns are categorical.
The InvoiceDate column is incorrectly stored as an object and requires conversion to datetime format.
Missing values were identified in the Description and CustomerID columns, with CustomerID having a significant number of missing entries.

In [18]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [19]:
# Drop rows with missing Description
df = df.dropna(subset=['Description'])

# Check result
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133626
Country             0
dtype: int64

The Description column contained 1,454 missing values, which were removed because product descriptions are essential for analysis.
The CustomerID column had a large number of missing values (135,080), which were retained since they likely represent guest transactions and removing them would significantly reduce the dataset.

After removing rows with missing product descriptions, the dataset no longer contains null values in the Description column. The number of missing CustomerID values slightly decreased as some of those rows were also removed. CustomerID missing values were retained to preserve transaction data.

In [20]:
#Check for duplicates
df.duplicated().sum()

5268

In [21]:
# Remove duplicates
df = df.drop_duplicates()

# Check again
df.duplicated().sum()

0

The dataset contained 5,268 duplicate records, which were removed to ensure data accuracy and prevent inflated analysis results. Duplicate entries can distort metrics such as total sales and product demand.

In [36]:
#Convert InvoiceDate to Datetime
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

# Check
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 524878 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    524878 non-null  object        
 1   StockCode    524878 non-null  object        
 2   Description  524878 non-null  object        
 3   Quantity     524878 non-null  int64         
 4   InvoiceDate  524878 non-null  datetime64[ns]
 5   UnitPrice    524878 non-null  float64       
 6   CustomerID   392692 non-null  float64       
 7   Country      524878 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 36.0+ MB


In [37]:
#Check negative quantity
df[df['Quantity'] < 0].shape

(0, 8)

In [38]:
# Keep only positive quantities
df = df[df['Quantity'] > 0]

# Check again
df[df['Quantity'] < 0].shape

(0, 8)

In [39]:
#Unit price check
df[df['UnitPrice'] <= 0].shape

(0, 8)

In [40]:
# Keep only valid prices
df = df[df['UnitPrice'] > 0]

# Check again
df[df['UnitPrice'] <= 0].shape

(0, 8)

The dataset contained 584 records with zero or negative unit prices. These were removed as they represent invalid or non-revenue-generating transactions and would negatively impact sales analysis.

In [43]:
df.to_csv("cleaned_dataset.csv", index=False)